dlt/destinations/impl/filesystem/filesystem.py

```diff
@@ -123,16 +124,51 @@ def write(self) -> None:
         from dlt.common.libs.deltalake import (
             write_delta_table,
             _deltalake_storage_options,
+            try_get_deltatable,
         )
 
         file_paths = [job.file_path for job in self.table_jobs]
 
-        write_delta_table(
-            path=self.client.make_remote_uri(self.make_remote_path()),
-            data=pa.dataset.dataset(file_paths),
-            write_disposition=self.table["write_disposition"],
-            storage_options=_deltalake_storage_options(self.client.config),
-        )
+        if (
+            self.table["write_disposition"] == "merge"
+            and (
+                dt := try_get_deltatable(
+                    self.client.make_remote_uri(self.make_remote_path()),
+                    storage_options=_deltalake_storage_options(self.client.config),
+                )
+            )
+            is not None
+        ):
+            assert self.table["x-merge-strategy"] in self.client.capabilities.supported_merge_strategies  # type: ignore[typeddict-item]
+
+            if self.table["x-merge-strategy"] == "upsert":  # type: ignore[typeddict-item]
+                if "parent" in self.table:
+                    unique_column = get_first_column_name_with_prop(self.table, "unique")
+                    predicate = f"target.{unique_column} = source.{unique_column}"
+                else:
+                    primary_keys = get_columns_names_with_prop(self.table, "primary_key")
```

Comment:
we don't use the merge keys? or did you overlook this? we should have failing tests if you don't and the tests are set up correctly.

Reply:
The `upsert` merge strategy does not support merge keys. It requires a primary key.

This [discussion](https://github.com/dlt-hub/dlt/pull/1294#issuecomment-2082372128) explains why.